### PR TITLE
feat: introduce logStreamActor & add notebook logs endpoint [DET-4100]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

- define helpers around handling logs for our APIs
- define and use LogEntry message for the API response
- add logStreamActor to handle streaming for logs.
- add notebook streaming log endpoint

## Test Plan

- create a notebook and listen for logs. 
- make sure the request is closed when the notebook is terminated
- added and ran test.sh

## Checklist

- [ ] remove temporary test files
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

**Comment by hamidzr** [comment]
@stoksc  this is already a 400 loc change; what are thoughts on not applying the same pattern to the 3 other entities (tensoboard, shells, commands) in this PR

File: master/test.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+id=$1
```
Comment (hamidzr) [inline master/test.sh:1]: temp: remove

File: master/internal/api_notebook.go [hunk 2]
```diff
@@ -27,3 +32,57 @@ func (a *apiServer) KillNotebook(
 	_ context.Context, req *apiv1.KillNotebookRequest) (resp *apiv1.KillNotebookResponse, err error) {
 	return resp, a.actorRequest(fmt.Sprintf("/notebooks/%s", req.NotebookId), req, &resp)
 }
+
+/* Command Helpers */
+// TODO reorder args?
+// CHECK parts of this code could live on the command and events actors but the idea is to
+// keep the api code out of there.
+func fetchCommandLogs(
```
Comment (hamidzr) [inline master/internal/api_notebook.go:40]: probably move to api/logs.go?

**Review by rb-determined-ai — COMMENTED** [review]
Ok, some high-level feedback as the first round of feedback.  I'm starting with the presupposition that we like the actor system we have and that we should generally stick to it.

Command logs are fundamentally tied to the actor system right now, but this PR does streaming as a goroutine.  I think things will be cleaner and just generally better if we make the thing that does streaming (processLogs) into an actor rather than just a goroutine that interacts with the actor system in an odd way (this ActorRequest object).

I imagine that every grpc request that comes in is its own goroutine, and that the grpc server expects you to fully respond to the response before returning from that goroutine (correct me if I'm wrong).  Basically, what I am suggesting is that the NotebookLogs creates an actor and then just blocks until that actor stops running (at least in the streaming case, in the non-streaming case this might be unnecessary).  That actor would be responsible for:

- registering with the eventManager actor as a subscriber to new logs
- registering with the command actor for a message when that command shuts down
- checking on the status of the grpc context
- passing logs through to the grpc context
- unregistering if the grpc context indicates that the connection is closed

This will have the following benefits:

- Easy event-based programming
- More consistent with the actor system
- If we use the same log structs in go (to imitate the same log type we plan on offering in the new api) then we can easily reuse this same actor type for trial logs in the future

What do you think?

PS, I bugged danny about the shape we expect for trial logs, and they did discuss this a bit in the [trial logs project ERD](https://docs.google.com/document/d/1P6U8X8X3ZUgB1S2878bsAZrJ7PnNhhBe3y7RDIkEo9o/edit#heading=h.owo8n2okijgw)

**Comment by hamidzr** [comment]
Thanks for the quick review! the current solution has been affected by the constraint that we'd like to wrap up the external parts of our new API sooner rather than later so external users, eg webui, cli, and other consumers, can start using it sooner rather than later and that has meant avoiding introducing new (and at times better) solutions where we have existing ones. For example, we're filtering and sorting objects in memory after fetching full responses from the database following existing patterns instead of building the necessary support delegating that to the database. It's not ideal but has let us to move forward and the external portion of the API out.

With that said using the event-based nature of the actor system wherever we can sounds like a great idea. I think this makes sense with streaming being the main and only target. I initially started by setting up the event manager to pass along events https://github.com/determined-ai/determined/pull/1248/commits/4ab0538525225ec40ca772dc260db0428d0a85ab but after chatting with @stoksc  I simplified it to what it is right now.

To add to the list of responsibilities you mention for the actor:


> That actor would be responsible for:

and  
- would also need to be able to get historical events as well
- support common request params: offset and limit

setting up actors as you describe it and writing support for allowing history to be requested through the actor system makes sense. I think the timing might be a question I'd need to figure out with @justin-determined-ai help.

Also part of the reason for setting up the request to the event manager actor this way (requesting batches) is to allow  us to support a non-streaming version of the endpoints  which won't be necessary if we the streaming endpoints can give us good throughput (more context: https://determined-ai.slack.com/archives/C013B86TBPC/p1597255589283800)

question: would we be passing individual log messages between the eventmanager actor and the notebook actor? we'll probably need to look at the overhead we might have there. :thinking:

**Comment by rb-determined-ai** [comment]
re: responsibilities: I think the historical logs and the offset management should be handled by the thing that stores up logs, and that thing is the already-existing eventManager actor.  Not the new streaming actor.  That's part of why you have that race condition that you called out in the code; right now the thing which stores up logs (eventManager) is different than the thing which calculates the offsets (processLogs goroutine).

re: individual message overhead: We already pass individual logs messages in memory all over the system.  Literally that's what the actor system is all about; everything is event based.  We will do buffering in special cases like to batch up writing to the database.  Or in other words, if the actor system has too much overhead, we're toast : )

At the end of the day, if we know the right way to implement something and it's not that much extra work, why not just do it?  If we have been talking about having streaming endpoints for how long... why not just do it?  Who will complain about great streaming endpoints?  And we are not that far from having them.

**Comment by hamidzr** [comment]
@rb-determined-ai  does https://github.com/determined-ai/determined/pull/1248/commits/d9982a7e341dfff85eb94b57700f786130e16115 look accurate to you in terms of what needs to be done to go the actor route? any changes

**Comment by rb-determined-ai** [comment]
I think if you look at https://github.com/determined-ai/determined/commit/d9982a7e341dfff85eb94b57700f786130e16115 you will see my comments with the code around them for context.

File: master/test-commands.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash -x
```
Comment (hamidzr) [inline master/test-commands.sh:1]: temporary. remove before merging.

**Comment by hamidzr** [comment]
@rb-determined-ai  I rewrote how we stream logs over using by standing up another actor. please take a look and lmk what suggestions you have.

File: master/internal/command/events.go [hunk 4]
```diff
@@ -44,17 +48,70 @@ type event struct {
...
+func (e *eventManager) RemoveSusbscribers(ctx *actor.Context) {
+	for actor := range e.logStreams {
+		// OPT this will trigger a bunch of CloseStream message that'll come back to eventManager.
+		ctx.Tell(actor, webAPI.CloseStream{})
+	}
+	// REMOVE ME
+	e.logStreams = make(logSubscribers)
+}
+
+// TODO inline this?
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:99]: I prefer not inlining, I think it makes testing harder in general.

File: master/internal/command/events.go [hunk 5]
```diff
@@ -44,17 +48,70 @@ type event struct {
...
+func (e *eventManager) ProcessNewLogEvent(ctx *actor.Context, msg event) {
+	// Publish.
+	for streamActor, logRequest := range e.logStreams {
+		// OPT we could probably use actor hierarchy to message multiple logStreamActors at once and say
+		// utilize multicast if we were over a capable network..
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:104]: I'd guess we are probably years away from running multiple masters, where this might be a concern.

File: master/internal/command/events.go [hunk 6]
```diff
@@ -44,17 +48,70 @@ type event struct {
...
+			count++
+		}
+	})
+	return count
+}
+
+// TODO inline this.
+func (e *eventManager) RemoveSusbscribers(ctx *actor.Context) {
+	for actor := range e.logStreams {
+		// OPT this will trigger a bunch of CloseStream message that'll come back to eventManager.
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:92]: That's fine.  This is a fundamental race condition when you have objects with unstructured lifetimes (structured = one starts before and ends after the other).  Either could try to close the other in either order, and they need to be able to tolerate that.

The [GObject memory model](https://developer.gnome.org/gobject/stable/gobject-memory.html) (part of glib, from the GNOME team) deals with this by explicitly separating object shutdown into two phases: "disposal" and "finalizing".  Disposal must tolerate being called many times, as two objects might try and trigger each other's disposal functions *from within* each other's disposal functions.  But "finalizing" is guaranteed to only happen once, after nobody holds a reference to you anymore.

In the actor system, this means when the actor is closed, all `Tell` and `Ask` operations will still complete, but nothing is actually added to the inbox.  That deals with tolerating the undefined shutdown order.  The finalization step is when golang GC's the actual actor inbox, which happens when nobody holds a reference to that actor anymore.

TL;DR: this comment isn't necessary haha

File: master/internal/command/events.go [hunk 7]
```diff
@@ -44,17 +48,70 @@ type event struct {
...
+		ctx.Tell(actor, webAPI.CloseStream{})
+	}
+	// REMOVE ME
+	e.logStreams = make(logSubscribers)
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:96]: What's this for?  Could you do `e.logStreams = nil`?

File: master/internal/command/events.go [hunk 8]
```diff
@@ -44,17 +48,70 @@ type event struct {
...
-	buffer     *ring.Ring
-	closed     bool
-	seq        int
+	bufferSize   int
+	buffer       *ring.Ring
+	closed       bool
+	seq          int
+	isTerminated bool // DISCUSS If we don't want to don't keep track of this here we'd either need to
+	// expose event externally or define and send a new message. also the logActor might not receive
+	// the termination log message (based on LogRequest) if we rely on exposing Event..
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:64]: Why would the logActor not receive the termination log message?

If it is paging through some middle part of the log, that's fine, isTerminated wouldn't matter.

If it is in a follow, then the log actor would either receive the terminated event, or a CloseStream message as the eventlog hit PostStop, right?

File: master/internal/command/events.go [hunk 9]
```diff
@@ -150,6 +245,57 @@ func validEvent(e event, greaterThanSeq, lessThanSeq *int) bool {
...
+	case ev.ExitedEvent != nil:
+		message = fmt.Sprintf("%s was terminated: %s", description, *ev.ExitedEvent)
+	case ev.LogEvent != nil:
+		message = fmt.Sprintf(*ev.LogEvent)
+	default:
+		// We rely on log entry IDs to provide pagination and since some of these events aren't actually
+		// log events we'd need to notify of them about these non existing logs either by adding a new
+		// attribute to our response or a sentient log entry or we could keep it simple and normalize
+		// command events as log struct by setting a special message.
+		// return nil, errors.New(fmt.Sprintf("event %v has no supported log message", ev))
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:267]: Question: if the pagination is handled in the same place as where the logs are buffered up (that is, in the eventManager), then can't we get rid of this detail?

File: proto/src/determined/log/v1/log.proto [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+package determined.log.v1;
+option go_package = "github.com/determined-ai/determined/proto/pkg/logv1";
+
+import "protoc-gen-swagger/options/annotations.proto";
+
+// LogEntry is a log event.
+message LogEntry {
+    option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+        json_schema: {
+            required: ["id", "message"]
```
Comment (rb-determined-ai) [inline proto/src/determined/log/v1/log.proto:12]: doesn't requiring `message` break the detail about having some non-message logs?

File: master/internal/api/actor.go [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"reflect"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/determined-ai/determined/master/pkg/actor"
+)
+
+// ActorRequest is a helper to ask an actor and populate the response in v.
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/api/actor.go:12]: it's not needed anymore :+1:

[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply master/internal/api/actor.go:12]: the original one too?  Or this publically-exposed one?

File: master/internal/command/events.go [hunk 12]
```diff
@@ -89,6 +94,21 @@ func (e *eventManager) Receive(ctx *actor.Context) error {
 			}
 		}
 
+	case GetEventCount:
+		// OPT we could use log_buffer here instead of a plain ring buffer.
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/command/events.go]: log_buffer.go  
it specifically works with a log entry (defined there) but

1. keeps track of how many actualentries have been written
2. can grab slices of logs back based on typical filters without scanning over them.

File: master/internal/api/actor.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/api/actor.go:12]: just this one.

File: master/internal/api/logs.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+// TerminationCheck checks whether a log processing should stop or not.
+type TerminationCheck func() (bool, error)
+
+// LogEntryToProtoLogEntry turns a logger.LogEntry into logv1.LogEntry.
+func LogEntryToProtoLogEntry(logEntry *logger.Entry) *logv1.LogEntry {
+	return &logv1.LogEntry{Id: int32(logEntry.ID), Message: logEntry.Message}
+}
+
+// ProcessLogs handles fetching and processing logs from a log store.
+func ProcessLogs(ctx context.Context,
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/api/logs.go:41]: I rewrote the streaming master logs using this. we could keep it and maybe later set non-streaming trial logs to use it as well (before the trial logs project) :thinking: 
however this in not really related to this PR

File: master/internal/api_notebook.go [hunk 14]
```diff
@@ -27,3 +34,46 @@ func (a *apiServer) KillNotebook(
...
+		api.NewCommandLogStreamActor(
+			resp.Context(),
+			eventManager,
+			logRequest,
+			onLogEntry,
+		),
+	)
+
+	if !created {
+		// Either there is a collision in actor address or actor creation failed.
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/api_notebook.go:71]: hmm maybe a better place for it would be the ActorOf function itself. it isn't called out what that bool is supposed to represent (there is a named var assigned to it at on code branch but..).

File: master/internal/api/logs.go [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+	ctx          context.Context
+	send         OnLogEntry
+	eventManager *actor.Ref
+}
+
+// CloseStream indicates that the log stream should close.
+type CloseStream struct{}
+
+// NewCommandLogStreamActor creates a new command logStreamActor.
+func NewCommandLogStreamActor(
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/api/logs.go:98]: that'd be nice I did think about this a little. I don't remember why I didn't proceed with it. I'll look into this.

File: master/internal/command/events.go (hunk 7 above)
Comment (hamidzr) [reply master/internal/command/events.go:96]: huh, didn't know nil would still be a valid slice :+1:

File: master/internal/command/events.go (hunk 8 above)
Comment (hamidzr) [reply master/internal/command/events.go:64]: > If it is paging through some middle part of the log, that's fine, isTerminated wouldn't matter.

right I think I got confused on this part

File: proto/src/determined/log/v1/log.proto (hunk 10 above)
Comment (hamidzr) [reply proto/src/determined/log/v1/log.proto:12]: this is just annotation but still not marking it is better

File: master/internal/command/events.go (hunk 9 above)
Comment (hamidzr) [reply master/internal/command/events.go:267]: hmm, the goal is to clarify why we basically send, at this point, empty log messages back so that the client can also keep track

File: master/internal/api_notebook.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/api_notebook.go:71]: i removed and commented on the return value of ActorOf

File: master/internal/api/logs.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/api/logs.go:98]: done

File: master/internal/command/events.go (hunk 6 above)
Comment (hamidzr) [reply master/internal/command/events.go:92]: thanks for the detailed explanation!

File: master/internal/api_trials.go [hunk 16]
```diff
@@ -35,6 +36,7 @@ func trialStatus(d *db.PgDB, trialID int32) (model.State, int, error) {
 	return trialStatus.State, trialStatus.NumLogs, err
 }
 
+// CHECK migrate to api.ProcessLogs?
```
Comment (hamidzr) [inline master/internal/api_trials.go:39]: remove

File: master/internal/command/events.go [hunk 17]
```diff
@@ -44,17 +48,63 @@ type event struct {
...
-	bufferSize int
-	buffer     *ring.Ring
-	closed     bool
-	seq        int
+	bufferSize   int
+	buffer       *ring.Ring
+	closed       bool
+	seq          int
+	isTerminated bool // DISCUSS If we don't want to don't keep track of this here we'd either need to
+	// expose event externally or define and send a new message.
```
Comment (hamidzr) [inline master/internal/command/events.go:63]: DISCUSS If we don't want to don't keep track of this here we'd either need to
expose event externally or define and send a new message.

File: master/internal/command/events.go [hunk 18]
```diff
@@ -88,6 +136,42 @@ func (e *eventManager) Receive(ctx *actor.Context) error {
...
+		}
+		delete(e.logStreams, ctx.Sender())
+
+	case webAPI.LogsRequest:
+		if ctx.Sender() == nil {
+			return errors.New(ctxMissingSender)
+		}
+		ctx.Respond(true)
+
+		total := countNonNullRingValues(e.buffer)
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:153]: First, why would we have a null value in the buffer at all?

Second, is there *any* way we can avoid this O(N) iteration through the whole buffer for every single LogsRequest?

File: master/internal/api/logs.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply master/internal/api/logs.go:41]: Question: Why not rewrite the streaming master logs using the logstream actor you just wrote instead?

File: master/internal/command/events.go (hunk 18 above)
Comment (hamidzr) [reply master/internal/command/events.go:153]: > First, why would we have a null value in the buffer at all?

I'll check again but I think that's how it's already initialized.. maybe we can change that.  

> Second, is there any way we can avoid this O(N) iteration through the whole buffer for every single LogsRequest?

the log_buffer keeps track of a counter.. or if we didn't null values
￼

**Comment by hamidzr** [comment]
> Question: Why not rewrite the streaming master logs using the logstream actor you just wrote instead?

chatted with @rb-determined-ai: since master logs are not already backed by an actor we'll leave this for later.

File: master/internal/command/events.go (hunk 18 above)
Comment (hamidzr) [reply master/internal/command/events.go:153]: created https://determinedai.atlassian.net/browse/DET-4206 for this. please add more details if it's missing anything.

**Comment by hamidzr** [comment]
I added a section to track how many log events are sent to the user so we can enforce the limit in the follow=true case.

File: master/internal/command/command.go [hunk 19]
```diff
@@ -123,6 +123,9 @@ func (c *command) Receive(ctx *actor.Context) error {
 	case scheduler.ResourcesAllocated, scheduler.ReleaseResources:
 		return c.receiveSchedulerMsg(ctx)
 
+		// TODO propagate command kill to subscribed logActors
+		// OR just to the event manager always?
+		// message to terminate logActor
```
Comment (rb-determined-ai) [inline master/internal/command/command.go:128]: Didn't we decide on how to handle this?

File: master/internal/command/events.go [hunk 20]
```diff
@@ -2,22 +2,37 @@ package command
...
 	"github.com/determined-ai/determined/master/pkg/actor/api"
 	"github.com/determined-ai/determined/master/pkg/check"
+	"github.com/determined-ai/determined/master/pkg/logger"
 )
 
 const defaultEventBufferSize = 200
+const ctxMissingSender = "message is missing sender information"
+
+func countNonNullRingValues(ring *ring.Ring) int {
+	// OPTIMIZE if this proves to be slow we could add a constant time solution.
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:27]: Make this a TODO and reference the ticket you made

File: master/internal/command/events.go [hunk 21]
```diff
@@ -88,6 +135,43 @@ func (e *eventManager) Receive(ctx *actor.Context) error {
...
+
+		if msg.Follow && !e.isTerminated && !limitMet {
+			e.logStreams[ctx.Sender()] = msg
+		} else {
+			ctx.Tell(ctx.Sender(), webAPI.CloseStream{})
+		}
+
+	case webAPI.CloseStream:
+		if ctx.Sender() == nil {
+			return errors.New(ctxMissingSender)
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:169]: make this a panic, after the infra ag meeting where we decided to use panics on internal logic errors.

**Review by rb-determined-ai — COMMENTED** [review]
Lookin real good!

File: master/internal/command/command.go (hunk 19 above)
Comment (hamidzr) [reply master/internal/command/command.go:128]: yeah, this is done. how did I leave this in :facepalm: I had to do a rebase at the end

File: master/internal/api_master.go [hunk 22]
```diff
@@ -30,37 +29,17 @@ func (a *apiServer) GetMaster(
...
-func effectiveOffset(reqOffset int, total int) (offset int) {
-	switch {
-	case reqOffset < -total:
-		return 0
-	case reqOffset < 0:
-		return total + reqOffset
-	default:
-		return reqOffset
-	}
+// TODO update the response format to wrap logv1.LogEntry.
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/api_master.go:32]: I just addressed it.

File: master/internal/command/events.go (hunk 21 above)
Comment (hamidzr) [reply master/internal/command/events.go]: what type of errors should we panic on ? maybe also places you'd usually do sanity checks?

File: master/internal/command/command.go (hunk 19 above)
Comment (hamidzr) [reply master/internal/command/command.go]: also https://github.com/determined-ai/determined/pull/1248#discussion_r491152479

File: master/internal/command/events.go (hunk 21 above)
Comment (rb-determined-ai) [reply master/internal/command/events.go]: We should panic on errors which are 100% internal logic failures.  Like the kind of errors where "this should literally never occur unless there's a bug in the system".

File: master/internal/command/command.go (hunk 19 above)
Comment (rb-determined-ai) [reply master/internal/command/command.go]: FWIW I find one big commit to be like a million times better when rebasing.

File: master/internal/api_notebook.go [hunk 23]
```diff
@@ -27,3 +36,49 @@ func (a *apiServer) KillNotebook(
...
+		),
+	)
+
+	if !created {
+		return errors.New("failed to create actor")
+	}
+
+	// We delegate checking for context closure to logStreamActor accepting that
+	// it could stay up if there are no log messages coming in instead of busy checking
+	// to kill the actor.
```
Comment (rb-determined-ai) [inline master/internal/api_notebook.go:82]: I don't think this comment is really necessary.  If you'd like to keep it, let's shorten it to "keep the GRPC context open until the logStream actor stops" or something?

File: master/internal/command/events.go [hunk 24]
```diff
@@ -15,9 +16,23 @@ import (
...
 	"github.com/determined-ai/determined/master/pkg/actor/api"
 	"github.com/determined-ai/determined/master/pkg/check"
+	"github.com/determined-ai/determined/master/pkg/logger"
 )
 
 const defaultEventBufferSize = 200
+const ctxMissingSender = "message is missing sender information"
+
+func countNonNullRingValues(ring *ring.Ring) int {
+	// TODO we could add a constant time solution.
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:26]: Change to the TODO format we normally use in the harness and master:
```
// TODO(DET-3262): ...
```

File: master/internal/command/events.go [hunk 25]
```diff
@@ -88,6 +135,43 @@ func (e *eventManager) Receive(ctx *actor.Context) error {
...
+		if ctx.Sender() == nil {
+			panic(ctxMissingSender)
+		}
+		ctx.Respond(true)
+
+		total := countNonNullRingValues(e.buffer)
+		offset := webAPI.EffectiveOffset(msg.Offset, total)
+		msg.Offset = offset
+
+		// Stream existing matching entries.
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:150]: Not sure this comment adds anything

File: master/internal/command/events.go [hunk 26]
```diff
@@ -44,17 +59,49 @@ type event struct {
...
+func (e *eventManager) processNewLogEvent(ctx *actor.Context, msg event) {
+	// Publish.
+	for streamActor, logRequest := range e.logStreams {
+		if eventSatisfiesLogRequest(logRequest, &msg) {
+			entry := eventToLogEntry(&msg)
+			ctx.Tell(streamActor, *entry)
+		}
+	}
+
+	// Remove terminated subscribers.
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:101]: ditto

File: master/internal/command/events.go [hunk 27]
```diff
@@ -44,17 +59,49 @@ type event struct {
...
+
+func (e *eventManager) removeSusbscribers(ctx *actor.Context) {
+	for actor := range e.logStreams {
+		ctx.Tell(actor, webAPI.CloseStream{})
+	}
+	e.logStreams = nil
+}
+
+func (e *eventManager) processNewLogEvent(ctx *actor.Context, msg event) {
+	// Publish.
```
Comment (rb-determined-ai) [inline master/internal/command/events.go:93]: ditto

File: master/internal/api/params.go [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+	case reqOffset < 0:
+		return int(math.Max(float64(total+reqOffset), 0))
+	default:
+		return reqOffset
+	}
+}
+
+// EffectiveLimit returns effective limit.
+// Input: non-negative offset and limit.
+func EffectiveLimit(limit int, offset int, total int) int {
```
Comment (rb-determined-ai) [inline master/internal/api/params.go:19]: panic if offset is negative?

It appears that limit is allowed to be negative right?  It's not clear if that's true from the comment.

File: master/internal/api/params.go [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+	case reqOffset < -total:
+		return 0
+	case reqOffset < 0:
+		return int(math.Max(float64(total+reqOffset), 0))
+	default:
+		return reqOffset
+	}
+}
+
+// EffectiveLimit returns effective limit.
```
Comment (rb-determined-ai) [inline master/internal/api/params.go:17]: I understand that this comment is required for exported functions, but if we have to have it, maybe explain what "effective" means?

File: master/internal/api/params.go [hunk 30]
```diff
@@ -0,0 +1,35 @@
+package api
+
+import "math"
+
+// EffectiveOffset returns effective offset.
```
Comment (rb-determined-ai) [inline master/internal/api/params.go:5]: ditto.

File: master/internal/api/logs.go [hunk 31]
```diff
@@ -0,0 +1,139 @@
...
+func (l *LogStreamActor) Receive(ctx *actor.Context) error {
+	switch msg := ctx.Message().(type) {
+	case actor.PreStart:
+		if response := ctx.Ask(l.logStore, l.req); response.Empty() {
+			ctx.Self().Stop()
+			return status.Errorf(codes.NotFound, "logStore did not respond")
+		}
+
+	case logger.Entry:
+		// Make sure the context is still open.
```
Comment (rb-determined-ai) [inline master/internal/api/logs.go:116]: prolly not necessary

File: master/internal/api/logs.go [hunk 32]
```diff
@@ -0,0 +1,139 @@
...
+
+// LogsRequest describes the parameters needed to target a subset of logs.
+type LogsRequest struct {
+	Offset int
+	Limit  int
+	Follow bool
+}
+
+// OnLogEntry is a callback called on each log entry.
+type OnLogEntry func(*logger.Entry) error
```
Comment (rb-determined-ai) [inline master/internal/api/logs.go:27]: Either make `OnLogEntry` into `OnLogEntryFn` or remove the `Fn` from `LogFetcherFn` (I'd prefer the first)

File: master/internal/api/logs.go [hunk 33]
```diff
@@ -0,0 +1,139 @@
...
+
+// LogFetcherFn fetchs returns a subset of logs based on a LogRequest.
+type LogFetcherFn func(LogsRequest) ([]*logger.Entry, error)
+
+// TerminationCheck checks whether a log processing should stop or not.
+type TerminationCheck func() (bool, error)
```
Comment (rb-determined-ai) [inline master/internal/api/logs.go:33]: ditto

**Review by rb-determined-ai — APPROVED** [review]
lgtm

File: master/internal/command/command.go (hunk 19 above)
Comment (hamidzr) [reply master/internal/command/command.go]: > FWIW I find one big commit to be like a million times better when rebasing.

I agree, I sometimes do mash them into one if I have to go through a big rebase. When I'm reviewing I'd like to see separate commits pushed after my review with no rebase so I can exactly see what was changed.

File: master/internal/api/params.go (hunk 28 above)
Comment (hamidzr) [reply master/internal/api/params.go:19]: I addressed both.
